pkg/pmem-csi-driver/nodeserver.go

```diff
@@ -88,57 +101,92 @@ func (ns *nodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
 	if len(req.GetTargetPath()) == 0 {
 		return nil, status.Error(codes.InvalidArgument, "Target path missing in request")
 	}
-	if len(req.GetStagingTargetPath()) == 0 {
-		return nil, status.Error(codes.InvalidArgument, "Staging target path missing in request")
-	}
 
 	// Serialize by VolumeId
 	volumeMutex.LockKey(req.GetVolumeId())
 	defer volumeMutex.UnlockKey(req.GetVolumeId())
 
-	targetPath := req.TargetPath
-	stagingtargetPath := req.StagingTargetPath
-	// TODO: check is bind-mount already made
-	// (happens when publish is asked repeatedly for already published namespace)
-	notMnt, err := mount.New("").IsLikelyNotMountPoint(targetPath)
-	if err != nil && !os.IsNotExist(err) {
-		return nil, status.Error(codes.Internal, "validate target path: "+err.Error())
-	}
-	if !notMnt {
-		// TODO(https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/95): check if mount is compatible. Return OK if it is, or appropriate error.
-		/*
-			1) Target Path MUST be the vol referenced by vol ID
-			2) VolumeCapability MUST match
-			3) Readonly MUST match
-		*/
-		return &csi.NodePublishVolumeResponse{}, nil
-	}
+	srcPath := req.GetStagingTargetPath()
+	ephemeral := false
 
-	if err := os.Mkdir(targetPath, os.FileMode(0755)); err != nil {
-		// Kubernetes is violating the CSI spec and creates the
-		// directory for us
-		// (https://github.com/kubernetes/kubernetes/issues/75535). We
-		// allow that by ignoring the "already exists" error.
-		if !os.IsExist(err) {
-			return nil, status.Error(codes.Internal, "make target dir: "+err.Error())
+	// Kubernetes v1.16+ would request ephemeral volumes via VolumeContext
+	val, ok := req.GetVolumeContext()[volumeContextEphemeral]
+	if ok {
+		b, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
 		}
+		ephemeral = b
+	} else {
+		// If met all below conditions we still treat as a ephemeral volume request:
+		// 1) No Volume found with given volume id
+		// 2) No provisioner info found in VolumeContext "storage.kubernetes.io/csiProvisionerIdentity"
+		// 3) No StagingPath in the request
+		device, _ := ns.cs.dm.GetDevice(req.VolumeId)
```

Comment:
> > devicemanager: Make GetDevice() return pointer
> 
> Why?

Its easier/cleaner to deal with pointers than values for a struct type, especially while check if a device is valid using `nil`

Reply:
> This ignores the error from `GetDevice`. Wouldn't it be cleaner to introduce a "not found" error type and check for that?

Yes, I agree to introduce error codes. But in this case, the only error returned(by both implementations) is  "not found" so we are not missing/ignoring any other errors.

Reply:
So here the caller is allowed to ignore the error code because the only error is "not found", which isn't documented. That looks rather fragile to me - what if we introduce a second error later? Are you sure that this call site will then be updated?

If you don't want to check the error here and continue to test for nil, then better remove the error return from `GetDevice` entirely. Then it is obvious from the function signature that the function always succeeds (returns the device) or returns nil when not found.

Reply:
GetDevice() now returns os.ErrExist, and the same check added here.